pkg/sources/gitlab/gitlab.go

```diff
@@ -564,3 +606,77 @@ func normalizeRepos(repos []string) ([]string, []error) {
 	}
 	return validRepos, errs
 }
+
+// Enumerate reports all GitLab repositories to be scanned to the reporter. If
+// none are configured, it will find all repositories within all projects that
+// the configured user has access to, while respecting the configured ignore
+// rules.
+func (s *Source) Enumerate(ctx context.Context, reporter sources.UnitReporter) error {
+	// Start client.
+	apiClient, err := s.newClient()
+	if err != nil {
+		return err
+	}
+
+	// Get repos within target.
+	repos, errs := normalizeRepos(s.repos)
+	for _, repoErr := range errs {
+		ctx.Logger().Info("error normalizing repo", "error", repoErr)
+		if err := reporter.UnitErr(ctx, repoErr); err != nil {
+			return err
+		}
+	}
+
+	// End early if we had errors getting specified repos but none were validated.
+	if len(errs) > 0 && len(repos) == 0 {
+		return fmt.Errorf("all configured repos had validation issues")
+	}
+
+	// Report all repos if specified.
+	if len(repos) > 0 {
+		for _, repo := range repos {
+			unit := sources.CommonSourceUnit{ID: repo}
+			if err := reporter.UnitOk(ctx, unit); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// Otherwise, enumerate all repos.
+	ignoreRepo := buildIgnorer(s.ignoreRepos, func(err error, pattern string) {
+		ctx.Logger().Error(err, "could not compile ignore repo glob", "glob", pattern)
+		// TODO: Handle error returned from UnitErr.
```

Comment:
Any reason you can't just return it? An un-compilable ignore glob sounds pretty fatal to me.

Reply:
It's in a callback right now, so it's just adding the error to the report. This comment is referring to the error when adding the error to the report, which can happen from context cancellation. If that's the case, it'll throw the same error when returning later in this function, so I don't think it's worth the effort to escape the error(s) out of the closure.

But you bring up a good point about un-compilable globs - should we make it fatal? Right now it appears they are not.

Reply:
Oh, right, derp again. It's probably a good thing to handle at some point but I don't think you need to do it as part of this PR. (I believe it's written that way so that `Validate` can detect it, so it's not like it's completely undiscoverable.)